miniapps/shifted_methods/distance.cpp

```diff
@@ -0,0 +1,315 @@
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.Parse();
+   if (!args.Good())
+   {
+      if (myid == 0)
+      {
+         args.PrintUsage(cout);
+      }
+      MPI_Finalize();
+      return 1;
+   }
+   if (myid == 0) { args.PrintOptions(cout); }
+
+   // 3. Enable hardware devices such as GPUs, and programming models such as
+   //    CUDA, OCCA, RAJA and OpenMP based on command line options.
+   Device device(device_config);
+   if (myid == 0) { device.Print(); }
+
+   // Refine the mesh.
+   Mesh mesh(mesh_file, 1, 1);
+   const int dim = mesh.Dimension();
+   for (int lev = 0; lev < rs_levels; lev++) { mesh.UniformRefinement(); }
+
+   // MPI distribution.
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   mesh.Clear();
+
+   Coefficient *ls_coeff;
+   int smooth_steps;
+   if (problem == 0)
+   {
+      ls_coeff = new DeltaCoefficient(0.5, -0.5, 1000.0);
+      smooth_steps = 0;
+   }
+   else if (problem == 1)
+   {
+      ls_coeff = new FunctionCoefficient(sphere_ls);
+      smooth_steps = 5;
+   }
+   else if (problem == 2)
+   {
+      ls_coeff = new FunctionCoefficient(sine_ls);
+      smooth_steps = 1;
+   }
+   else
+   {
+      ls_coeff = new FunctionCoefficient(Gyroid);
+      smooth_steps = 0;
+   }
+
+   DistanceSolver *dist_solver = NULL;
+   if (solver_type == 0)
+   {
+      auto ds = new HeatDistanceSolver(t_param);
+      if (problem == 0)
+      {
+         ds->transform = false;
+         ds->diffuse_iter = 1;
+      }
+      ds->smooth_steps = smooth_steps;
+      ds->vis_glvis = false;
+      dist_solver = ds;
+   }
+   else if (solver_type == 1)
+   {
+      const int p = 10;
+      const int newton_iter = 50;
+      auto ds = new PLapDistanceSolver(p, newton_iter);
+      dist_solver = ds;
+   }
+   else { MFEM_ABORT("Wrong solver option."); }
+   dist_solver->print_level = 1;
+
+   H1_FECollection fec(order, dim);
+   ParFiniteElementSpace pfes_s(&pmesh, &fec), pfes_v(&pmesh, &fec, dim);
+   ParGridFunction distance_s(&pfes_s), distance_v(&pfes_v);
+   dist_solver->ComputeScalarDistance(*ls_coeff, distance_s);
+   dist_solver->ComputeVectorDistance(*ls_coeff, distance_v);
+
+   ParGridFunction input_ls(distance_s.ParFESpace());
+   input_ls.ProjectCoefficient(*ls_coeff);
+   delete ls_coeff;
+
+   // Send the solution by socket to a GLVis server.
+   if (visualization)
+   {
+      int size = 500;
+      char vishost[] = "localhost";
+      int  visport   = 19916;
+
+      socketstream sol_sock_w(vishost, visport);
+      sol_sock_w << "parallel " << num_procs << " " << myid << "\n";
+      sol_sock_w.precision(8);
+      sol_sock_w << "solution\n" << pmesh << input_ls;
+      sol_sock_w << "window_geometry " << 0 << " " << 0 << " "
+                 << size << " " << size << "\n"
+                 << "window_title '" << "Input Level Set" << "'\n" << flush;
+
```

Comment:
I have been getting consistent crashes in GLVis when I run the "Problem 1" sample run on the periodic-cube mesh. The problem (I believe) is related to multiple connection attempts from multiple processors arriving in the wrong sequence. I think the fix is to place `MPI_Barrier` calls between the socket stream calls.  See [ex5p.cpp](https://github.com/mfem/mfem/blob/b0770915511fc63fda3120b13c88db48946e4302/examples/ex5p.cpp#L460) for another code that uses the same trick.